1822 character interface with llm rag web services

AKissmehl opened the pull request.

andy-kanta committed:
The first draft of the backend servers, has issues. Don't merge!

andy-kanta committed:
still work in progress!

andy-kanta committed:
- Finish rewrite the code
- Cleaned up some death code
- Adapting the SaveLoadAudioUtilities.cs with the AudioClipToByteArray methode.

andy-kanta committed:
Remove a node.

andy-kanta committed:
Update based on feedback.

AKissmehl committed:
Merge branch 'develop' into 1822-character-interface-with-llm-rag-web-services

fwild reviewed (changes requested):
The config should come from assets/resources/miragexrconfig.txt - maybe an opportunity to change that to JSON as well!

andy-kanta committed:
Work in progress. 2 knowen bugs.

andy-kanta committed:
Merge remote-tracking branch 'origin/1822-character-interface-with-llm-rag-web-services' into 1822-character-interface-with-llm-rag-web-services

# Conflicts:
#	Assets/MirageXR/Player/Scripts/RootObject.cs
#	Assets/XR/Settings/Open XR Package Settings.asset
#	Packages/packages-lock.json

andy-kanta committed:
Merge remote-tracking branch 'origin/develop' into 1822-character-interface-with-llm-rag-web-services

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+        /// <returns>A String with the result of task or an error if an network error appears</returns>
+        public async Task<string> Listen(AudioClip audioClip, string model)
+        {
+            UnityEngine.Debug.LogError("Speak");
```
only for debug?

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+        /// <returns>A boolean. True if the operations was successful, falls if it wasn't.</returns>
+        private async Task<bool> GetOptions()
+        {
+            UnityEngine.Debug.LogError("GetOptions");
```
remove

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+        /// <returns><placeholder>A <see cref="Tasks.Task"/> representing the asynchronous operation.</placeholder></returns>
+        public async Task<AudioClip> Speak(string speakOut, string voice, string model)
+        {
+            UnityEngine.Debug.LogError("Speak");
```
remove

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+            var r = new AiServicesConfig(apiURL, username, password);
+            r.SetToken(token);
+            _config = r;
+            UnityEngine.Debug.LogError("Config is done with " + _config.AiToken + _config.AiApiUrl + _config.AiPassword + _config.AiUsername);
```
remove

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+                    $"Token = {token}");
+            }
+            UnityEngine.Debug.LogError(apiURL+ username+ password);
+            var r = new AiServicesConfig(apiURL, username, password);
```
```suggestion
            var config = new AiServicesConfig(apiURL, username, password);
```

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+            {
+                throw new HttpRequestException($"Error while receiving the token: {request.error}!");
+            }
+            var r = request.downloadHandler.text;
```
```suggestion
            var text = request.downloadHandler.text;
```

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+        /// </summary>
+        private readonly struct TokenResponse // still needed? @todo..
+        {
+            public TokenResponse(string t)
```
```suggestion
            public TokenResponse(string token)
```

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+        /// <param name="user"> Username</param>
+        /// <param name="pass"> Password</param>
+        /// <returns>The token as struct</returns>
+        private static async Task<string> AuthenticateUser(string apiURL, string user, string pass)
```
```suggestion
        private static async Task<string> AuthenticateUser(string apiURL, string user, string password)
```

cawke commented on Assets/StreamingAssets/MirageXRConfig.txt:
```diff
@@ -0,0 +1,26 @@
+companyName:WEKIT ECS
```
is this an old file?

cawke commented on Assets/Resources/MirageXRConfig.txt:
```diff
@@ -0,0 +1,17 @@
+companyName:WEKIT ECS
```
is this an old file?

cawke commented on Assets/StreamingAssets/config.json:
```diff
@@ -0,0 +1,23 @@
+{
```
is this an old file?

cawke commented on Assets/Resources/MirageXRConfig.json:
```diff
@@ -0,0 +1,22 @@
+  "calibrationMarker":"Assets/Editor/Vuforia/ImageTargetTextures/MirageXRTargets/target_scaled.jpg",
+  "pdfCalibrationMarker":"Assets/MirageXR/Config/Resources/CalibrationImage/MirageXR_calibration_image_pdf.bytes",
+  "AiApiUrl": "http://127.0.0.1:8000/",
+  "AI_USERNAME": "admin" ,
```
you need to put your login password into a separate file. we cannot store private data in a public repository

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,334 @@
+
+namespace MirageXR
+{
+    public class AiServices
```
you wrote a service. now you need to write a manager that will initialize it (read the config and receive the login password) and manage this service.

AKissmehl replied:
Okay, can you point me to an existing manager (MirageXR/Managers/?) and explain to me what it should do? I Podbay have to use the unity method like start() awake() and so on. Can we have a chat about what the manager should do?

cawke replied:
Now your AiServices class combines the work of both manager and service. in our case service should provide convenient access to work with your server (for example, your methods "Listen", "Think", "Speak" and "AuthenticateUser", although I would change the names). The manager should prepare (initialize) this service: read the keys (AiApiUrl, AiUsername, AiPassword) (we can't store these keys in the public repository, so put them in a file in the "Resources" folder), create an instance of the service, get and store the authentication token. Also all work with the service should be done through the manager, which means that in the manager you need to add methods "Listen", "Think", "Speak" inside which you will check the availability of the token, call the methods of the service and handle possible errors.

example https://github.com/WEKIT-ECS/MIRAGE-XR/blob/release_2.6/Assets/MirageXR/Scripts/OpenAI/OpenAIManager.cs

cawke commented on Assets/MirageXR/Player/Scripts/RootObject.cs:
```diff
@@ -24,6 +24,7 @@ public class RootObject : MonoBehaviour
         private MoodleManager _moodleManager;
         private EditorSceneService _editorSceneService;
         private WorkplaceManager _workplaceManager;
+        private AiServices _aiServices;
```
there must be a manager who will manage this service

cawke commented on Assets/MirageXR/ContentTypes/Character/Characters/ImageDisplayFiles/MagicParticle.mat:
```diff
@@ -2,14 +2,15 @@
 %TAG !u! tag:unity3d.com,2011:
```
I think this change was made by mistake. roll it back

cawke commented on Assets/AddressableAssetsData/AssetGroups/Glyphs.asset:
```diff
@@ -133,6 +133,10 @@ MonoBehaviour:
     m_Address: Glyphs/Unscrew
     m_ReadOnly: 0
     m_SerializedLabels: []
+  - m_GUID: cea501bde7c9ce54dacac44c0cc2e01e
```
I think this change was made by mistake. roll it back

cawke commented on Packages/packages-lock.json:
```diff
@@ -31,13 +31,14 @@
     "com.microsoft.mixedreality.openxr": {
```
I think this change was made by mistake. roll it back

cawke commented on Assets/XR/Settings/Open XR Package Settings.asset:
```diff
@@ -1,6 +1,6 @@
 %YAML 1.1
 %TAG !u! tag:unity3d.com,2011:
---- !u!114 &-9152892435972783649
```
I think this change was made by mistake. roll it back

andy-kanta committed:
Request fixes. Issue with a STT request. Work in progress.

cawke commented on Assets/MirageXR/Scripts/AiServices/AiServices.cs:
```diff
@@ -0,0 +1,479 @@
+            {
+                new MultipartFormDataSection("model", model),
+                //new MultipartFormFileSection("audio", audioBytArray),
+                new MultipartFormDataSection("dfsdfjk", "dsfsdfkhsdfksdfh"),
```
I'm guessing it's the remnants of your testing?

andy-kanta committed:
Refactor AiServices methods and improve code readability

The 'AiServices' methods were updated: 'Listen', 'Think', 'Speak' and 'GetOptions' were renamed to 'ConvertSpeechToTextAsync', 'SendMessageToLlm', 'ConvertTextToSpeechAsync', and 'GetAvailableModels' respectively, for better clarity. Adjustments were also made to the implementation details, such as reducing duplicated checks, removing unnecessary comments and unused imports, enhancing the header setting in the 'ConvertTextToSpeechAsync' method, and adjusting the data form used when making the 'UnityWebRequest'.

andy-kanta committed:
Rollbacks of changes that were done by mistake.

cawke committed:
small fix

cawke committed:
ai manager has been added

AKissmehl committed:
Merge pull request #1974 from WEKIT-ECS/ashubin/1822-ai-manager

ai manager for "1822 character interface"

andy-kanta committed:
rollback.

AKissmehl committed:
Delete Assets/Resources/MirageXRConfig.json

andy-kanta committed:
rollback

andy-kanta committed:
Merge remote-tracking branch 'origin/1822-character-interface-with-llm-rag-web-services' into 1822-character-interface-with-llm-rag-web-services

AKissmehl committed:
Update CharacterAugmentation.cs

rollback

andy-kanta committed:
The AIManager and AI services have been refactored with a focus on how the Endpoint object  will be deserialized.  A class that olds the Token and an AI Model has been added. The AI model will be held in a List for each endpoint and can be reached over a get method.

Old configuration files have been removed, and an error log has been added if the Authentication fails.

Open to discuss:
public async Task<List<OptionsResponse>> GetModelsAsync()
 is, from my point of view, obsolete and can be removed.

andy-kanta committed:
Merge remote-tracking branch 'origin/1822-character-interface-with-llm-rag-web-services' into 1822-character-interface-with-llm-rag-web-services

AKissmehl commented:
@cawke 
I did not follow the result of our discussion 1 to 1, since I found a simpler solution (at least I think) which needs just an AI Model class and not 3 specific ones. That is the description for the last commit:
The AIManager and AI services have been refactored with a focus on how the Endpoint object will be deserialized.  A class that olds the Token and an AI Model has been added. The AI model will be held in a List for each endpoint and can be reached over a get method.

Old configuration files have been removed, and an error log has been added if the Authentication fails.

Open to discuss:
public async Task<List<OptionsResponse>> GetModelsAsync()
 is, from my point of view, obsolete and can be removed.
 
 I have one Issue, the 11
[Assets/MixedRealityToolkit.Generated/ProjectPreferences.asset](https://github.com/WEKIT-ECS/MIRAGE-XR/pull/1911/files#diff-eaddd58c6f4b03814db0457cb5785ebee3c3f724799ea69d5456b69465712422) is a copy past version out of develop. It still shows up in the PR and I don't know how to remove it.

cawke commented on Assets/MirageXR/Managers/AIManager/AIManager.cs:
```diff
@@ -0,0 +1,194 @@
+            }
+            UnityEngine.Debug.Log("Set models done! LLM:"+_llmModels.Count+" TTS:"+_ttsModels.Count+ "STT:"+_sttModels.Count );
+
+        }
```
```suggestion
        private void SetModels(List<AIModel> endpointModels)
        {
            foreach (var model in endpointModels)
            {
                AddModelBasedOnEndpointName(model);
            }
        }
```

cawke commented on Assets/MirageXR/Managers/AIManager/AIManager.cs:
```diff
@@ -0,0 +1,194 @@
+                return null;
+            }
+        }
+         */
```
```suggestion
```

cawke commented on Assets/MirageXR/Managers/AIManager/AIManagerDataModel/Token.cs:
```diff
@@ -0,0 +1,17 @@
+            BackendToken = text;
+        }
+    }
+}
```
```suggestion
namespace MirageXR.AIManagerDataModel
{
    public class Token
    {
         [JsonProperty("token")] public string Token { get; set; }
    }
}
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+            {
+                UnityEngine.Debug.LogError(
+                    $"Error while receiving the result of the Think endpoint: {webRequest.error}");
+            }
```
```suggestion
            {
                throw new HttpRequestException(
                    $"Error while receiving the result of the Think endpoint: {webRequest.error}");
            }
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+                UnityEngine.Debug.LogError("Error fetching or deserializing data: " + e.Message);
+                return null;
+            }
+        }
```
```suggestion
            catch (Exception e)
            {
                throw new HttpRequestException("Error fetching or deserializing data: " + e.Message);
            }
        }
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+            if (webRequest.result != UnityWebRequest.Result.Success)
+            {
+                UnityEngine.Debug.LogError($"Error while receiving the result of the Speak endpoint: {webRequest.error} {webRequest.result}");
+            }
```
```suggestion
            if (webRequest.result != UnityWebRequest.Result.Success)
            {
                throw new HttpRequestException($"Error while receiving the result of the Speak endpoint: {webRequest.error} {webRequest.result}");
            }
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+            await request.SendWebRequest();
+            if (request.result != UnityWebRequest.Result.Success)
+            {
+                UnityEngine.Debug.LogError($"Error while receiving the token: {request.error}!");
```
```suggestion
                throw new HttpRequestException($"Error while receiving the token: {request.error}!");
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+
+            text = text.Substring(10, text.Length - 10 - 2);
+            Token token = new Token(text);
+            return token;
```
```suggestion
            try
            {
                var token = JsonConvert.DeserializeObject<TokenData>(optionsText);
                return token.Token;
            }
            catch (Exception e)
            {
                throw new HttpRequestException("Error fetching or deserializing data: " + e.Message);
            }
```

cawke commented on Assets/MirageXR/Managers/AIManager/AiServices.cs:
```diff
@@ -0,0 +1,232 @@
+        /// <param name="user">Username</param>
+        /// <param name="password">Password</param>
+        /// <returns>The authentication token</returns>
+        public static async Task<Token> AuthenticateUserAsync(string apiURL, string user, string password)
```
```suggestion
        public static async Task<string> AuthenticateUserAsync(string apiURL, string user, string password)
```

cawke commented on Assets/MirageXR/Managers/AIManager/AIManager.cs:
```diff
@@ -0,0 +1,194 @@
+        private string _username;
+        private string _password;
+
+        private Token _token;
```
```suggestion
        private string _token;
```

andy-kanta committed:
Update bes on the Feedback and add comments for Documentation.
